search/utils.hpp

```diff
@@ -118,4 +120,31 @@ void ForEachCategoryTypeFuzzy(StringSliceBase const & slice, TLocales const & lo
     });
   }
 }
+
+// Returns whether the request specified by |slice| is categorial
+// in any of the |locales|. We expect that categorial requests should
+// mostly arise from clicking on a category button in the UI.
+// It is assumed that typing a word that matches a category's name
+// and a space after it means that no errors were made.
+template <typename T>
+bool IsCategorialRequest(QuerySliceOnRawStrings<T> const & slice, Locales const & locales,
+                         CategoriesHolder const & catHolder)
+{
+  if (slice.Size() != 1 || slice.HasPrefixToken())
+    return false;
+
+  bool found = false;
+  auto token = slice.Get(0);
+  catHolder.ForEachName([&](CategoriesHolder::Category::Name const & categorySynonym) {
+    if (std::find(locales.begin(), locales.end(), categorySynonym.m_locale) == locales.end())
```

Comment:
nit: you can use SmallSet for locales.

Reply:
I like the idea, but let's do it in a separate PR, and for all places that use Locales at once.